exonum/src/runtime/rust/service.rs

```diff
@@ -53,21 +68,57 @@ pub trait Service: ServiceDispatcher + Debug + 'static {
         Ok(())
     }
 
+    /// Return a list of root hashes of all Merkelized tables defined by this service,
+    /// as of the given snapshot of the blockchain state and service instance descriptor.
+
+    /// The core uses this list to [aggregate][1] hashes of tables defined by all services into a
+    /// single Merkelized meta-map.
+    /// The hash of this meta-map is considered the hash of the entire blockchain [state][2] and
+    /// is recorded as such in blocks and Precommit messages.
+    ///
+    /// [See also.][3]
+    ///
+    /// [1]: ../struct.StateHashAggregator.html
+    /// [2]: ../../blockchain/struct.Block.html#structfield.state_hash
+    /// [3]: ../../blockchain/struct.Schema.html#method.state_hash_aggregator
     fn state_hash(&self, instance: InstanceDescriptor, snapshot: &dyn Snapshot) -> Vec<Hash>;
 
+    /// Perform storage operations on behalf of the service before committing the block.
+    ///
+    /// This changes affect `state_hash` that means this changes must be same for each node.
+    /// In other words you should use only data from context.
+    ///
+    /// The order of invoking `before_commit` method for every service instance depends on the ID.
+    /// `before_commit` for the service instance with the smallest ID is invoked first up to the
+    /// largest one.
+    /// Effectively, this means that services should not rely on a particular ordering of
+    /// Service::execute invocations.
```

Comment:
```suggestion
    /// `Service::execute` invocations.
```

shouldn't it be `Service::before_commit` btw?

Reply:
Oh, this is copypaste from old doc ^_^